lib/async_hooks.js

```diff
@@ -190,6 +191,123 @@ class AsyncResource {
 }
 
 
+// AsyncLocal //
+
+const kStack = Symbol('stack');
+const kIsFirst = Symbol('is-first');
+const kMap = Symbol('map');
+const kOnChangedCb = Symbol('on-changed-cb');
+const kHooks = Symbol('hooks');
+const kSet = Symbol('set');
+
+class AsyncLocal {
+  constructor(options = {}) {
+    if (typeof options !== 'object' || options === null)
+      throw new ERR_INVALID_ARG_TYPE('options', 'Object', options);
+
+    const { onChangedCb = null } = options;
+    if (onChangedCb !== null && typeof onChangedCb !== 'function')
+      throw new ERR_INVALID_ARG_TYPE('options.onChangedCb',
+                                     'function',
+                                     onChangedCb);
+
+    this[kOnChangedCb] = onChangedCb;
+    this[kMap] = new Map();
+
+    const fns = {
+      init: (asyncId, type, triggerAsyncId, resource) => {
+        // Propagate value from current id to new (execution graph)
+        const value = this[kMap].get(executionAsyncId());
+        if (value)
+          this[kMap].set(asyncId, value);
+      },
+
+      destroy: (asyncId) => this[kSet](asyncId, null),
+    };
+
+    if (this[kOnChangedCb]) {
+      // Change notification requires to keep a stack of async local values
+      this[kStack] = [];
+      // Indicates that first value was stored (before callback "missing")
+      this[kIsFirst] = true;
+
+      // Use before/after hooks to signal changes because of execution
+      fns.before = (asyncId) => {
+        const stack = this[kStack];
+        const cVal = this[kMap].get(asyncId);
+        const pVal = stack[stack.length - 1];
+        stack.push(pVal);
```

Comment:
Should this be `stack.push(cVal)`?

Reply:
no, I have to remember the previous value in `before` as I needed it as current value in `after`. the current value in `before` is available via the map.